Unit tests for some env option functions

nrwahl2 opened the pull request:
Closes [T493](lib/common/tests/options/Makefile.am).

clumens commented on lib/common/tests/options/pcmk__env_option_test.c:
```diff
@@ -0,0 +1,152 @@
+#include <stdlib.h>
+#include <string.h>
+#include <setjmp.h>
+#include <cmocka.h>
```
Once #2798 gets merged, you can reduce much of this block to `#include <crm/common/unittest_internal.h>` - see [8d8dbd4](https://github.com/ClusterLabs/pacemaker/pull/2798/commits/8d8dbd466325f836f91b6bfd262c727cb1191144).

clumens commented on lib/common/tests/options/Makefile.am:
```diff
@@ -0,0 +1,21 @@
+AM_CPPFLAGS = -I$(top_srcdir)/include		\
+	      -I$(top_builddir)/include		\
+	      -I$(top_srcdir)/lib/common
+LDADD = $(top_builddir)/lib/common/libcrmcommon_test.la
```
Once #2798 gets merged, you'll want to add `AM_CFLAGS = -DPCMK__UNIT_TESTING` right here - see [54c0620](https://github.com/ClusterLabs/pacemaker/pull/2798/commits/54c062060f120db94bbe25b2dc3d10bb6af40302).  You'll also want to add `-lcmocka` to `LDADD` - see [4dcd34a](https://github.com/ClusterLabs/pacemaker/pull/2798/commits/4dcd34a05b2bf8a7c6608167d7e064529f4e02a9).

Also, I feel like these Makefile.am files can probably get reduced a lot more.  I think these blocks at the top are just about identical at this point.

clumens reviewed (commented):
Once #2798 gets merged, there's a couple things you should change here for consistency - nothing major.

nrwahl2 commented on lib/common/mock.c:
```diff
@@ -70,15 +70,22 @@ __wrap_calloc(size_t nmemb, size_t size)
 {
-    return pcmk__mock_getenv? mock_ptr_type(char *) : __real_getenv(name);
+    if (!pcmk__mock_getenv) {
+        __real_getenv(name);
```
Should be `return`. Will fix.

nrwahl2 replied:
(making a note that no regression test catches this; does `mock.c` even get pulled into a regular build?)

clumens replied:
No, mock.c is only built into libcrmcommon_test as part of "make check".

nrwahl2 commented on lib/common/options.c:
```diff
@@ -337,7 +355,7 @@ pcmk__set_env_option(const char *option, const char *value)
  * daemon.
  *
- * \param[in] daemon   Daemon name
+ * \param[in] daemon   Daemon name (can be NULL)
```
A NULL `daemon` is currently allowed, even though we're not calling this with a NULL `daemon` now. If `daemon` is NULL and `value` is true, we still return true. So might as well document it.

But we do need to NULL-check `daemon` before passing it to `strstr()`.

nrwahl2 commented on lib/common/options.c:
```diff
@@ -289,8 +290,16 @@ pcmk__env_option(const char *option)
-        snprintf(env_name, NAME_MAX, "%s%s", prefixes[i], option);
+        int rv = snprintf(env_name, NAME_MAX, "%s%s", prefixes[i], option);
+
+        if (rv >= sizeof(env_name)) {
```
I considered also checking for a negative `rv` just in case. Still open to it. It's reasonable, but the logical conclusion is to add a check and error handler everywhere we call `snprintf()`... and there are a lot of them.

kgaillot replied:
Yes actually, that is a goal. Taking care of these is worthwhile but don't worry about the rest. I plan to replace the offset+= calls with GString or something along those lines.

nrwahl2 replied:
Done.

nrwahl2 commented on lib/common/options.c:
```diff
@@ -317,9 +326,18 @@ pcmk__set_env_option(const char *option, const char *value)
-        if (value) {
+        int rv = snprintf(env_name, NAME_MAX, "%s%s", prefixes[i], option);
+
+        if (rv >= sizeof(env_name)) {
```
I feel like `pcmk__set_env_option()` should return a boolean, indicating whether or not it successfully both variables. I didn't worry about it for this PR.

kgaillot replied:
bool should be reserved for true/false semantics; standard Pacemaker return codes should be used for success/failure, even if only two codes are possible.

Since it's an internal function, whether this one should have a return value should depend mainly on whether the callers should do anything differently based on success or failure.

nrwahl2 replied:
> return codes

Makes sense.

> do anything

I would expect at least some of them *should*, if we fail to set the env options that we're trying to set. Maybe, maybe not.

kgaillot replied:
It would be worthwhile to log errors inside the function itself, but since these tend to be for the benefit of resource agents, there's not much else to do

nrwahl2 commented on lib/common/tests/options/pcmk__env_option_enabled_test.c:
```diff
@@ -0,0 +1,130 @@
+#include "mock_private.h"
+
+static void
+disabled_null_value(void **state)
```
I wonder if we should mock `pcmk__env_option()`. So far only system calls are mocked. But this test basically tests `pcmk__env_option()` again via `pcmk__env_option_enabled()`, which is redundant.

All we need to test here is whether, **given** a NULL return value from `pcmk__env_option()`, `pcmk__env_option_enabled()` returns false. We don't care what **causes** `pcmk__env_option()` to return NULL.

kgaillot replied:
No need to mock pcmk__env_option(), nor repeat all possible tests for it or crm_is_true(). We just want to test that pcmk__env_option_enabled() is combining the two correctly.

nrwahl2 replied:
Cool, so it sounds like I could just pick any given `option` (e.g., `NULL`) that will cause `pcmk__env_option()` to return `NULL`, in order to test this.

nrwahl2 replied:
Done.

nrwahl2 commented on lib/common/options.c:
```diff
@@ -347,7 +365,9 @@ pcmk__env_option_enabled(const char *daemon, const char *option)
-    return (value != NULL) && (crm_is_true(value) || strstr(value, daemon));
+    return (value != NULL)
+        && (crm_is_true(value)
+            || ((daemon != NULL) && (strstr(value, daemon) != NULL)));
```
I'd like to add a regex check at some point, rather than a simple substring check. Allow comma or whitespace delimiters but make sure the daemon exists as a discrete "word".

kgaillot replied:
See pcmk__strcmp(..., pcmk__str_regex)

nrwahl2 replied:
Thanks! I did already find that one, just didn't want to add more clutter to this PR :)

nrwahl2 commented:
Build failure is just fedora rawhide RPM

kgaillot commented:
The related PRs have been merged, this can be rebased and updated

nrwahl2 commented on mk/tap.mk:
```diff
@@ -22,7 +22,9 @@ WRAPPED = calloc		\
 	  getgrent		\
 	  getpwnam_r		\
 	  readlink		\
+	  setenv		\
```
This is aligned in the raw file

kgaillot commented:
@clumens , could you review this one when you get a chance? Thanks

clumens commented on lib/common/tests/options/pcmk__set_env_option_test.c:
```diff
@@ -0,0 +1,144 @@
+    /* pcmk__set_env_option() wants to set "PCMK_<option>" and "HA_<option>". If
+     * "PCMK_<option>" is too long for the buffer, it simply moves on to
+     * "HA_<option>". A string of length (NAME_MAX - 3) will set us just over
+     * just over the edge for both tries.
```
"just over just over"

nrwahl2 replied:
Whoops, thanks.

kgaillot commented on lib/common/options.c:
```diff
@@ -318,24 +314,18 @@ pcmk__env_option(const char *option)
-        unsetenv(env_name);
+    for (int i = 0; i < (sizeof(prefixes) / sizeof(char *)); i++) {
+        snprintf(env_name, NAME_MAX, "%s%s", prefixes[i], option);
+        if (value) {
```
BTW if you haven't seen that guideline yet, the intent is to improve readability by giving an immediate indication that the variable is a pointer (vs. a boolean or integer). When you're looking at the code later it avoids having to scroll back and forth to see the declaration or making the wrong assumption about it.

nrwahl2 replied:
I changed it to `if (value != NULL)` in the very next commit (the one for validation) :)

This first commit did nothing but deduplicate code.

nrwahl2 commented on lib/common/options.c:
```diff
@@ -285,21 +285,17 @@ pcmk__cli_help(char cmd, crm_exit_t exit_code)
-    if (value != NULL) {
-        crm_trace("Found %s = %s", env_name, value);
-        return value;
+    for (int i = 0; i < (sizeof(prefixes) / sizeof(char *)); i++) {
```
Oh neat

nrwahl2 commented on lib/common/mock.c:
```diff
@@ -67,29 +67,62 @@ __wrap_calloc(size_t nmemb, size_t size)
  */
 
-bool pcmk__mock_getenv = false;
+bool pcmk__mock_env = false;
```
I figured if we were mocking the getter, we'd usually want to mock the setter and vice-versa (or it at least wouldn't hurt to mock the other). And that we would want to keep the boolean enabled for as little time as possible anyway.

I'm happy to do separate variables for `pcmk__mock_getenv`, `_setenv`, and `_unsetenv`, if you'd prefer. Actually, that's what I did in my first pass at this. Then later I noticed that we use one variable `pcmk__mock_grent` for all the `*grent` calls. I revised this to use one variable accordingly.

kgaillot replied:
For the grent functions, they can't be used independently, much like opendir/readdir/closedir, so they have to be combined. These could be used independently, but maybe more likely not to, so I think it's fine either way.

nrwahl2 commented on lib/common/mock.c:
```diff
@@ -72,15 +72,22 @@ __wrap_calloc(size_t nmemb, size_t size)
+        return __real_getenv(name);
+    }
+    check_expected_ptr(name);
+    return mock_ptr_type(char *);
```
I don't honestly see why, as long as it's in a separate commit and is in service of the same goal, but I can make one

nrwahl2 replied:
@clumens Sorry, I misread and thought you were talking about putting the `mock.c` changes from this PR into a separate PR. I see what you mean now, about `expect_*`.

nrwahl2 replied:
I started exploring this. It seems that `CRM_ASSERT()`/`pcmk__assert_asserts()` breaks the CMocka expect/check sequence. I don't want to take us off on a tangent, just want to leave a note while it's on my mind.  Checking expectations may not be *quite* as straightforward for every mocked function.

With `expect_*` lines, CMocka acts as if `__wrap_calloc()` was never called.
```
    pcmk__mock_calloc = true;   // calloc() will return NULL
                                // (40 * 1.01) + 601
    expect_any(__wrap_calloc, nmemb);
    expect_any(__wrap_calloc, size);
    pcmk__assert_asserts(pcmk__compress(SIMPLE_DATA, 40, 0, &result, &len));
    pcmk__mock_calloc = false;  // Use the real calloc()

FAIL: pcmk__compress_test 3 - calloc_fails
# __wrap_calloc: '%s' parameter still has values that haven't been checked.
# pcmk__compress_test.c:49: note: remaining item was declared here
# '%s' parameter still has values that haven't been checked.
# pcmk__compress_test.c:50: note: remaining item was declared here
# not ok - tests
```

After removing the `expect_any` lines, CMocka acts as if `__wrap_calloc()` **was** called and didn't have any `expect`s on the stack.
```
FAIL: pcmk__compress_test 3 - calloc_fails
# No entries for symbol __wrap_calloc.
# mock.c:70: error: Could not get value to check parameter nmemb of function __wrap_calloc
# There were no previously declared parameter values for this test.
ERROR: expr terminated without asserting
```

nrwahl2 replied:
https://github.com/ClusterLabs/pacemaker/pull/2811

nrwahl2 commented on lib/common/mock.c:
```diff
@@ -67,29 +67,62 @@ __wrap_calloc(size_t nmemb, size_t size)
+    }
+    check_expected_ptr(name);
+    return 0;
+}
```
Probably so. My thought was to add stuff to the mocked functions on an as-needed basis. The functions that we were going to unit test with mocked `setenv()`/`unsetenv()` didn't check for errors from those system calls.

Now I'm adding those error checks to the validate commit when I push again (see https://github.com/ClusterLabs/pacemaker/pull/2801#discussion_r949286427). So it might be worth incorporating `errno` into the test somehow... even though it doesn't change the return value (`void`) of `pcmk__set_env_option()`. We'll only be printing an error message in `pcmk__set_env_option()`, in case of `(un)setenv()` failure.

nrwahl2 commented:
Updated to address review.
* Check `snprintf()` for negative return value and respond as follows (https://github.com/ClusterLabs/pacemaker/pull/2801#discussion_r949286427):
    * Log an error message.
    * Return `NULL` from `pcmk__env_option()`.
    * Return from `pcmk__set_env_option()`. This function should probably be updated to return a standard Pacemaker return code, and its callers (at least some) should check the return code. That's outside the scope of this PR.
* Mock `snprintf()` and include tests for the above cases.
* Simplify tests `disabled_null()` test for `pcmk__env_option_enabled()`. Only use one input condition that causes `pcmk__env_option()` to return `NULL`. (https://github.com/ClusterLabs/pacemaker/pull/2801#discussion_r949292623)
* Use `PCMK__NELEM()` in loop control. (https://github.com/ClusterLabs/pacemaker/pull/2801#discussion_r949339751)
* Use separate mock booleans for `getenv()`, `setenv()`, `unsetenv()`. (https://github.com/ClusterLabs/pacemaker/pull/2801#discussion_r949354714)
* Allow mocked `setenv()` and `unsetenv()` to set `errno` based on `will_return()`. (https://github.com/ClusterLabs/pacemaker/pull/2801#discussion_r949356975)

nrwahl2 commented:
retest this please

nrwahl2 commented:
Really strange test errors... in both Ubuntus and in F36 Power9:
* `snprintf_error()` in `pcmk__env_option_test.c` results in a call to `getenv()`, even though `pcmk__env_option()` returns before calling `getenv()` when `snprintf()` returns a negative value.
```
19:59:04 not ok 2 - snprintf_error
19:59:04 FAIL: pcmk__env_option_test 2 - snprintf_error
19:59:04 # No entries for symbol __wrap_getenv.
19:59:04 # mock.c:90: error: Could not get value to check parameter name of function __wrap_getenv
19:59:04 # There were no previously declared parameter values for this test.
```
* `snprintf_error()` in `pcmk__set_env_option_test.c` does not result in any call to `snprintf()`, even though `pcmk__set_env_option()` explicitly calls `snprintf()`.
```
19:59:04 FAIL: pcmk__set_env_option_test 2 - snprintf_error
19:59:04 # %s() has remaining non-returned values.
19:59:04 # pcmk__set_env_option_test.c:49: note: remaining item was declared here
19:59:04 # __wrap_snprintf: '%s' parameter still has values that haven't been checked.
19:59:04 # pcmk__set_env_option_test.c:46: note: remaining item was declared here
19:59:04 # '%s' parameter still has values that haven't been checked.
19:59:04 # pcmk__set_env_option_test.c:47: note: remaining item was declared here
19:59:04 # '%s' parameter still has values that haven't been checked.
19:59:04 # pcmk__set_env_option_test.c:48: note: remaining item was declared here
```

All the other distros work fine.

nrwahl2 commented:
I got my hands on F36 ppc64le. This change is what breaks the "`snprintf_error`" unit tests there:
https://fedoraproject.org/wiki/Changes/PPC64LE_Float128_Transition

Running `gdb` on a minimal reproducer:
```
Breakpoint 1, func (s=s@entry=0x10020020 "hello world") at test.c:5
5	    snprintf(buf, 16, "%s", s);
(gdb) s
0x00007ffff7d54a98 in ___ieee128_snprintf (s=0x7fffffffeac0 "\360\352\377\377\377\177", maxlen=16, format=0x10020010 "%s") at ../sysdeps/ieee754/ldbl-128ibm-compat/ieee128-snprintf.c:24
```

Special cases and hidden overrides like this may keep popping up for `snprintf()` in the future. Maybe use a different approach (e.g., macro-based instead of wrapper-based) for mocking stuff like Xprintf?

Note: We technically don't need to check the return value of `snprintf()` for `pcmk__*env_option_enabled`. It only returns an error if `n > INT_MAX`, which should never happen since `n` is a defined constant. Ideally we would check it anyway.

If you all want, I'll remove the `snprintf()` mocking code and the two unit tests that use it tomorrow, and push that forward to another day.

kgaillot commented on lib/common/mock.c:
```diff
@@ -92,6 +92,62 @@ __wrap_getenv(const char *name)
+    check_expected_ptr(value);
+    check_expected(overwrite);
+    errno = mock_type(int);
+    return (errno == 0)? 0 : -1;
```
This should probably call __real_setenv() for errno == 0 (ditto for unsetenv)

nrwahl2 replied:
I'm on the fence about that.

On the one hand, `__wrap_getenv()` never calls `__real_getenv()` if `pcmk__mock_getenv` is true. Instead it returns the mocked value we gave it via `will_return()`. So it's asymmetrical to have `__wrap_{un,}setenv()` call `__real_{un,}setenv()` if `pcmk__mock_{un,}setenv` is true, even if `errno == 0`. In other words, if the mocked `getenv()` doesn't touch the real environment, you'd think the mocked `setenv()` and `unsetenv()` wouldn't either. That's why I didn't let them produce any side effects besides setting `errno`.

The caller of set/unset also doesn't have to use the "real" results of setting an environment variable since we can mock the return value of `getenv()`.

On the other hand, `__wrap_fopen()` does call `__real_fopen()` if `errno == 0`. Which makes sense since the caller has to actually use the results -- the open file handle. Same with the proposed `__wrap_snprintf()` calling `vsnprintf()` -- we need something in that buffer.

kgaillot reviewed (commented):
Looks good to me once platform issues are sorted out

nrwahl2 commented on lib/common/Makefile.am:
```diff
@@ -110,7 +110,9 @@ libcrmcommon_test_la_LDFLAGS	= $(libcrmcommon_la_LDFLAGS) -rpath $(libdir) $(LDF
-libcrmcommon_test_la_CFLAGS	= $(libcrmcommon_la_CFLAGS) -DPCMK__UNIT_TESTING -fno-builtin -fno-inline
+libcrmcommon_test_la_CFLAGS	= $(libcrmcommon_la_CFLAGS) -DPCMK__UNIT_TESTING \
+					-fno-builtin -fno-inline \
+					-U_FORTIFY_SOURCE
```
Interesting. I'm not really sure what tests that refers to -- probably a lot has changed since 2016. I get the gist about flag combinations in replacements.

kgaillot replied:
T191 is strictly about configure.ac, so tests like AC_COMPILE_IFELSE()

nrwahl2 commented:
Removed the `-U_FORTIFY_SOURCE` and `__wrap_snprintf()` commits from this PR. Mocking `snprintf()` will require a different approach or a hack, so it's better addressed separately. I also removed the `snprintf_error()` test cases from the `pcmk__env_option()` and `pcmk__set_env_option()` unit test files. We can add those back when `snprintf()` is taken care of.

As of right now, `__wrap_setenv()` and `__wrap_unsetenv()` still don't call `__real_X`, based on my comment in https://github.com/ClusterLabs/pacemaker/pull/2801#discussion_r951846421 about symmetry and the lack of need to use the real environment (we have a mocked `getenv()` we can use). If either you or @clumens feels strongly about calling `__real_X`, I can easily update both `__wrap_X` functions to make that call.
